pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go

```diff
@@ -184,7 +184,7 @@ func (m *Stub) Stop() error {
 	return m.cleanup()
 }
 
-func (m *Stub) Watch(resourceName string) {
+func (m *Stub) Watch(resourceName string, autoregister bool) {
```

Comment:
iff we need to add this feature to the `stub`, I think it should be a global option. Perhaps we should add a configurable callback like `stubGetPreferredAllocFunc` to control the registration?
I believe it should _not_ be a Watch option because is supposed to always hold the same value for the lifecycle of the plugin. IOW, we don't expect this function to be called sometimes with true, sometimes with false. Should always be consistent.

Reply:
> iff we need to add this feature to the stub, I think it should be a global option. Perhaps we should add a configurable callback like stubGetPreferredAllocFunc to control the registration?

Are you referring to the use of `GetPreferredAllocation` device plugin endpoint here or is the name co-incidental?  IIUC, that allows device plugin to provide its own preference for device allocation. I am not sure how we use that to control registration.  

> I believe it should not be a Watch option because is supposed to always hold the same value for the lifecycle of the plugin. IOW, we don't expect this function to be called sometimes with true, sometimes with false. Should always be consistent.

I am not sure I fully understand this comment. The overall intent of watch here is to watch for changes to the unix socket as kubelet deletes them when it starts. Once the kubelet restart is detected,  the gRPC should always be restarted. What we are doing with the `autoregister` argument is facilitating the cases where the device plugin wants to take control of its own re-registration  (which it can do by  specifying autoregister= false and happens in case of sample device plugin when it has REGISTER_CONTROL_FILE specified as environment var). Typical flow would involve both start of gRPC server and registration of the plugin to kubelet (autoregister= true).

Reply:
Sorry for my extremely terse and hard to understand comments. Let me elaborate a bit

>Are you referring to the use of GetPreferredAllocation device plugin endpoint here or is the name co-incidental? IIUC, that allows device plugin to provide its own preference for device allocation. I don't think that can be used to control the registration process.

I'd experiment adding a callback which client code can use as hook point to control the registration process. It's probably easier for all of us if I sketch out some code. Something like this:

```
diff --git a/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go b/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go
index 1d226f72a85..8d0bd4b2f01 100644
--- a/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go
+++ b/pkg/kubelet/cm/devicemanager/plugin/v1beta1/stub.go
@@ -53,11 +53,20 @@ type Stub struct {
 	// getPreferredAllocFunc is used for handling getPreferredAllocation request
 	getPreferredAllocFunc stubGetPreferredAllocFunc
 
+	// registerControlFunc is used to gate the registration requests
+	registerControlFunc stubRegisterControlFunc
+
 	registrationStatus chan watcherapi.RegistrationStatus // for testing
 	endpoint           string                             // for testing
 
 }
 
+type stubRegisterControlFunc func(err error) bool
+
+func defaultRegisterControlFunc(err error) bool {
+	return true
+}
+
 // stubGetPreferredAllocFunc is the function called when a getPreferredAllocation request is received from Kubelet
 type stubGetPreferredAllocFunc func(r *pluginapi.PreferredAllocationRequest, devs map[string]pluginapi.Device) (*pluginapi.PreferredAllocationResponse, error)
 
@@ -90,6 +99,7 @@ func NewDevicePluginStub(devs []*pluginapi.Device, socket string, name string, p
 
 		allocFunc:             defaultAllocFunc,
 		getPreferredAllocFunc: defaultGetPreferredAllocFunc,
+		registerControlFunc:   defaultRegisterControlFunc,
 	}
 }
 
@@ -103,6 +113,44 @@ func (m *Stub) SetAllocFunc(f stubAllocFunc) {
 	m.allocFunc = f
 }
 
+// SetRegisterControlFunc sets registerControlFunc of the device plugin
+func (m *Stub) SetRegisterControlFunc(f registerControlFunc) {
+	m.registerControlFunc = f
+}
+
+func (m *Stub) Watch(resourceName string) {
+	for {
+		select {
+		case stop := <-m.stopWatcher:
+			if stop {
+				return
+			}
+		default:
+			_, statErr := os.Lstat(m.socket)
+			if statErr != nil {
+				// Socket file not found; restart server
+				klog.InfoS("Server endpoint not found", m.socket)
+				klog.InfoS("Most likely Kubelet restarted")
+				if err := m.Restart(); err != nil {
+					klog.ErrorS(err, "Unable to restart server")
+					panic(err)
+
+				}
+
+				// we should probably loop until the callback passes, or clarify the semantic
+				// from stub perspective of restarting without reconnecting
+				if ok := m.registerControlFunc(statErr); ok {
+					if err := m.Register(pluginapi.KubeletSocket, resourceName, pluginapi.DevicePluginPath); err != nil {
+						klog.ErrorS(err, "Unable to register to kubelet")
+						panic(err)
+					}
+				}
+			}
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
 // Start starts the gRPC server of the device plugin. Can only
 // be called once.
 func (m *Stub) Start() error {
```

The reason I'm proposing a new field in `Stub` object here is because, much like other callbacks, I expect the registration behavior to be consistent across calls of `Watch`. IOW, `Watch` should likely be called with always the same set of parameters, invariant wrt the calls. This hints the property should be aobject property, not a call parameter. Furthermore, this enables us to set defaults and enable consumers of this library which are fine with the default behavior (always register) to safely ignore this details.

Arguably, we don't really need a callback. We may use a simple object field like a boolean. Using a callback seems however more consistent with the rest of the code and enables the client code to inject delays for free, so it seems worth the little extra complexity.

Reply:
Thanks for the suggestion, I see your point now. I will play around with the code and see how I can factor in your suggestion.

Reply:
Absolutely. I'm totally fine with partially adopting my suggestion above. My main goal is to try to enable good defaults and don't necessarily expose client code to detail which may be irrelevant in some cases (e.g. ALWAYS requiring a parameter)